### PR TITLE
feat: serialize subscription fields

### PR DESCRIPTION
**Pull request by aht007**
[Jira Ticket](https://2u-internal.atlassian.net/browse/PON-3)(2u-Internal)

## Description

The new program subscription model “program_subscription“ which were added in this [PR](https://github.com/openedx/course-discovery/pull/3833) should be serialized into the Program serializer(s) and its fields should be synced in the Algolia connector for Algolia search indexing and Contentful data.

### CODE REVIEW

**Comment by aht007** [comment]
This is how the data will appear on our programs api (api/v1/programs/):

For a program with subscription eligible True
Screenshot 2023-03-24 at 11 21 03 AM

For a program with subscription eligible False or N/A

Screenshot 2023-03-24 at 11 21 07 AM

Also on Alogolia we will have the below new fields as a result of these changes:
Screenshot 2023-03-24 at 9 56 04 PM

And these facets will also be added:

Screenshot 2023-03-24 at 9 55 53 PM

File: course_discovery/apps/api/serializers.py [hunk 1]
```diff
@@ -2025,6 +2060,27 @@ def get_card_image_url(self, obj):
...
+            return data
+
+        subscription_eligible = subscription.get('subscription_eligible', False)
+        subscription_prices = subscription.get('prices', [])
+
+        # Add subscription properties to data object
+        data['subscription_eligible'] = subscription_eligible
+        data['subscription_prices'] = subscription_prices
+
+        return data
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/serializers.py]: Why is this needed? Can't this be kept under `subscription_eligible`?

File: course_discovery/apps/api/v1/tests/test_views/test_programs.py [hunk 2]
```diff
@@ -258,7 +258,7 @@ def test_filter_by_type(self):
         program_type_name = 'foo'
         program = ProgramFactory(type__name_t=program_type_name, partner=self.partner)
         url = self.list_path + '?type=' + program_type_name
-        self.assert_list_results(url, [program], 18)
+        self.assert_list_results(url, [program], 21)
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/v1/tests/test_views/test_programs.py:261]: 3 counts? Is it 3 because of new serializer? Can you double check if that leads to N+1 problem?

File: course_discovery/apps/course_metadata/index.py [hunk 3]
```diff
@@ -79,7 +79,8 @@ class EnglishProductIndex(BaseProductIndex):
     facet_fields = (('availability_level', 'availability'), ('subject_names', 'subject'), ('levels', 'level'),
                     ('active_languages', 'language'), ('product_type', 'product'), ('program_types', 'program_type'),
                     ('staff_slugs', 'staff'), ('product_allowed_in', 'allowed_in'),
-                    ('product_blocked_in', 'blocked_in'))
+                    ('product_blocked_in', 'blocked_in'), 'subscription_eligible',
+                    'subscription_prices')
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/index.py:83]: Is there a need to facet on prices? Eligibility makes sense. But price is a list of dict, not sure what benefits will be there.

**Review by DawoudSheraz — COMMENTED** [review]
I have left a few comments on different parts of this PR.

File: course_discovery/apps/api/serializers.py (hunk 1 above)
Comment (aht007) [reply course_discovery/apps/api/serializers.py]: The original structure of the data would have a subscription dictionary which contained the field nested in it. We just removed the subscription dictionary here and instead placed the fields at the top level so that parsing the data is easy.

File: course_discovery/apps/api/v1/tests/test_views/test_programs.py (hunk 2 above)
Comment (aht007) [reply course_discovery/apps/api/v1/tests/test_views/test_programs.py:261]: I don't think it is leading to N+1 problems. In our program serializer we are prefetching the program subscription which prefetches all the subscription prices which in turn prefetches the currencies so an overall increase of 3 in the count.

File: course_discovery/apps/course_metadata/index.py (hunk 3 above)
Comment (aht007) [reply course_discovery/apps/course_metadata/index.py:83]: Prices is a list of dict but I wanted to add it so if someone wants to segregate records based on the subscription currency then they should be able to. Price field is not of much interest here for us.

File: course_discovery/apps/api/serializers.py [hunk 4]
```diff
@@ -2025,6 +2060,27 @@ def get_card_image_url(self, obj):
...
+
+        subscription_eligible = subscription.get('subscription_eligible', None)
+        subscription_prices = subscription.get('prices', None)
+
+        # Add subscription properties to data object
+        data['subscription_eligible'] = subscription_eligible
+        data['subscription_prices'] = subscription_prices
+
+        return data
+
```
Comment (uzairr) [inline course_discovery/apps/api/serializers.py:2070]: i think this code could be refactored:
`
data['subscription_prices'] = None if not subscription else subscription.get('subscription_prices', None)
`
the above same pattern can be used to extract the eligibility. After that the function size will be reduced to just 3,4 lines

File: course_discovery/apps/api/serializers.py (hunk 1 above)
Comment (DawoudSheraz) [reply course_discovery/apps/api/serializers.py]: Ok.

File: course_discovery/apps/api/v1/tests/test_views/test_programs.py (hunk 2 above)
Comment (DawoudSheraz) [reply course_discovery/apps/api/v1/tests/test_views/test_programs.py:261]: Can you list down which 3 queries are executing? Cause if we are adding 3 queries for each program, it can lead to  N+1.

File: course_discovery/apps/course_metadata/index.py (hunk 3 above)
Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/index.py:83]: Ah, Ok. I personally don't see any value in adding this field as facet field. Maybe I am missing some context here.

File: course_discovery/apps/api/v1/tests/test_views/test_programs.py (hunk 2 above)
Comment (aht007) [reply course_discovery/apps/api/v1/tests/test_views/test_programs.py:261]: These are the queries that are getting added, plus there is one for fetching the currencies too which is also being prefetched. So it wouldn't do three queries for every single program but 3 queries overall for all programs.

Screenshot 2023-03-29 at 12 08 49 PM

**Review by DawoudSheraz — APPROVED** [review]

**Review by uzairr — APPROVED** [review]
